driver/driver.cpp

```diff
@@ -0,0 +1,185 @@
+#include "driver.h"
+#include "environment.h"
+#include "connection.h"
+#include "descriptor.h"
+#include "statement.h"
+#include "ini_defines.h"
+
+#include <chrono>
+
+Driver::Driver() noexcept {
+    setAttrSilent(SQL_ATTR_TRACE, (isYes(INI_TRACE_DEFAULT) ? SQL_OPT_TRACE_ON : SQL_OPT_TRACE_OFF));
+    setAttr<std::string>(SQL_ATTR_TRACEFILE, INI_TRACEFILE_DEFAULT);
+}
+
+Driver::~Driver() {
+}
```

Comment:
> `getDriver().unregisterDescendant(*this);`, which accesses already destroyed `Driver` instance, and `descendant` (which is Environment currently being destroyed)

`Driver` instance is not destroyed, but destroying, `descendants` member is still intact at that moment.

In your suggested code, while the destructor is executing, the `Driver` instance will be unable to resolve any descendant by handle, theoretically, if requested. So a basic assumption about registering will be broken at that moment.

Reply:
Well, Ok it is still destroy**ing**, but it is already part-destroy**ed**. And I think it is really a bad idea trying to poke `Driver` (either from outside or from inside of lib) while it is destroying. With my piece of code I'm just trying to minimize potential harm of later member re-arragements, `unregisterDescendant` refactoring, etc.

Anyhow, this thing looks really fragile and not future-proof.

Reply:
I see your concerns. Will adding just an explicit `environments.clear();` to otherwise empty destructor body make it more stable?
```
Driver::~Driver(){
    // Make sure these are destroyed before anything else.
    environments.clear();
}
```

Reply:
Pushed my this suggestion.